### PR TITLE
contrib: Support prereleases in release prep scripts

### PR DESCRIPTION
**Pull request by joestringer**
Add support for vX.Y.0-rcN and vX.Y.0-snapshotN prereleases into the
start-release.sh and submit-release.sh scripts.

### CODE REVIEW

**Comment by joestringer** [comment]
I've been testing these out to generate https://github.com/cilium/cilium/pull/17501 with a standard Cilium release process by just specifying the RC tag version. Will leave at draft until we go through the full process to see if there are missing pieces.

**Comment by joestringer** [comment]
These changes will currently support preparing a vX.Y.0-rc0 release, but if we run it for vX.Y.0-rc1 then it will perform a diff between the latest vX.Y-1 series release rather than putting together a changelog for the vX.Y.0-rc0..vX.Y.0-rc1 changes.

**Comment by joestringer** [comment]
I tested the first commit using instructions similar to those now updated in the .github issue template commit, and prepared v1.11.0-rc0 using those changes. The middle patches I just put together and I've done some initial testing locally to see whether they do roughly the correct thing, but we'll need to go through the -rc1 process to make sure that they work as intended all the way through. The issue template / docs updates, I've just assembled after the process so I think they capture everything that is needed but for those we would also need to follow the release process once more based on these instructions to iron out any bugs or missing steps.

File: contrib/release/start-release.sh [hunk 1]
```diff
@@ -18,6 +19,14 @@ usage() {
     logecho "--help     Print this help message"
 }
 
+# $1 - VERSION
+version_is_prerelease() {
+    if echo "$1" | grep -q "\(rc\)\|\(snapshot\)"; then
+        return 0
+    fi
+    return 1
+}
```
Comment (qmonnet) [inline contrib/release/start-release.sh]: Ok, note that you could stick to bash (and avoid spawning two processes), e.g.
```bash
version_is_prerelease() {
    case "$1" in
        *rc*|*snapshot*)
            return 0
            ;;
        *)
            return 1
            ;;
    esac
}
```

File: contrib/release/start-release.sh [hunk 2]
```diff
@@ -29,12 +38,12 @@ handle_args() {
         common::exit 0
     fi
 
-    if ! echo "$1" | grep -q "[0-9]\+\.[0-9]\+\.[0-9]\+"; then
+    if ! echo "$1" | grep -q "[0-9]\+\.[0-9]\+\.[0-9]\+\(-\(\(rc\)\|\(snapshot\)\)\(\.\)\?[0-9]\+\)\?$"; then
```
Comment (qmonnet) [inline contrib/release/start-release.sh]: Can't this use `$RELEASE_REGEX` that you introduce in common.sh in the next commit? They look identical.

**Review by qmonnet — APPROVED** [review]
Looks good as far as I can tell, with two minor comments you may feel free to ignore.

The git-foo and regexp magic in these files, plus my lack of experience on the release process, make any change rather hard to follow. I didn't see anything concerning, but in the end I'm mostly relying on your tests and experiments. (Nice to have some of the regexp handling factored under common.sh though.)

File: contrib/release/start-release.sh (hunk 1 above)
Comment (joestringer) [reply contrib/release/start-release.sh]: :+1: my apologies for making review difficult here, I've tended to come at these from a rather "write-only" perspective since these scripts are byproducts of the actual release process and I'm always thinking "one of these days we'll rewrite this in Go..". Your suggestion looks sensible, I will adopt it.

File: contrib/release/start-release.sh (hunk 2 above)
Comment (joestringer) [reply contrib/release/start-release.sh]: yep, I scanned through but looks like I missed this one.

File: contrib/release/submit-release.sh [hunk 3]
```diff
@@ -51,12 +50,21 @@ if $GENERATE_SUMMARY; then
     CHANGELOG=$SUMMARY
     SUMMARY="$RELEASE-pr-$(date --rfc-3339=date).txt"
     echo "Prepare for release $RELEASE" > $SUMMARY
-    tail -n+4 $CHANGELOG >> $SUMMARY
+    if [ "$BRANCH" = "master" ]; then
+        echo "" >> $SUMMARY
+        echo "See the included CHANGELOG.md for a full list of changes." >> $SUMMARY
```
Comment (joestringer) [inline contrib/release/submit-release.sh:55]: FWIW I set this like this because RCs almost always have a changelog too big for the GH issue template. It'd be nice to have a more robust way to detect this for the .0 release as well since we'll include full changelog there as well I think, but :shrug: this is good enough for now.

**Review by aanm — APPROVED** [review]

File: contrib/release/start-release.sh (hunk 1 above)
Comment (qmonnet) [reply contrib/release/start-release.sh]: I don't hold you responsible for the difficulty here, sorry if my message read differently :). I just meant that it's hard because it's bash scripts with fine manipulations of git branches and version number, and I find it hard to track mentally during the review (what version number variable has that `v` prefix, where does it get/lose it, what branch are we supposed to be on at this point in the script, ...). Go would likely make it easier to follow, but if the bash scripts work it's all good at the moment.
